fcm/fcm.go

```diff
@@ -0,0 +1,200 @@
+package fcm
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+import (
+	"github.com/google/cloud-print-connector/log"
+	"github.com/google/cloud-print-connector/notification"
+)
+
+const (
+	gcpFcmSubscribePath = "fcm/subscribe"
+)
+
+type FCM struct {
+	fcmServerBindURL string
+	cachedToken      string
+	tokenRefreshTime time.Time
+	clientID         string
+	proxyName        string
+	fcmTTLSecs       float64
+	FcmSubscribe     func(string) (interface{}, error)
+
+	notifications chan<- notification.PrinterNotification
+	dead          chan struct{}
+
+	quit chan struct{}
+}
+
+type Notification struct {
+	Data   [][]interface{}
+	Number int
+}
+
+type Data struct {
+	Notification string `json:"notification"`
+	Subtype      string `json:"subtype"`
+}
+
+type FcmMessage struct {
+	Category    string `json:"category"`
+	CollapseKey string `json:"collapse_key"`
+	Data        Data   `json:"data"`
+	From        string `json:"from"`
+	MessageID   string `json:"message_id"`
+	TimeToLive  int    `json:"time_to_live"`
+}
+
+func NewFCM(clientID string, proxyName string, fcmServerBindURL string, FcmSubscribe func(string) (interface{}, error), notifications chan<- notification.PrinterNotification) (*FCM, error) {
+	f := FCM{
+		fcmServerBindURL,
+		"",
+		time.Time{},
+		clientID,
+		proxyName,
+		0,
+		FcmSubscribe,
+		notifications,
+		make(chan struct{}),
+		make(chan struct{}),
+	}
+	return &f, nil
+}
+
+//  get token from GCP and connect to FCM.
+func (f *FCM) Init() {
+	iidToken := f.GetToken()
+	if err := f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit); err != nil {
+		for err != nil {
+			log.Errorf("FCM restart failed, will try again in 10s: %s", err)
+			time.Sleep(10 * time.Second)
+			err = f.ConnectToFcm(f.notifications, iidToken, f.dead, f.quit)
+		}
+		log.Info("FCM conversation restarted successfully")
+	}
+
+	go f.KeepFcmAlive()
+}
+
+// Quit terminates the FCM conversation so that new jobs stop arriving.
+func (f *FCM) Quit() {
+	// Signal to KeepFCMAlive.
+	close(f.quit)
+}
+
+// Fcm notification listener
+func (f *FCM) ConnectToFcm(fcmNotifications chan<- notification.PrinterNotification, iidToken string, dead chan<- struct{}, quit chan<- struct{}) error {
+	log.Debugf("Connecting to %s?token=%s", f.fcmServerBindURL, iidToken)
+	resp, err := http.Get(fmt.Sprintf("%s?token=%s", f.fcmServerBindURL, iidToken))
+	if err != nil {
+		// failed for ever no need to retry
+		quit <- struct{}{}
+		return err
+	}
+	if resp.StatusCode == 200 {
+		reader := bufio.NewReader(resp.Body)
+		go func() {
+			for {
+				line, err := reader.ReadBytes('\n')
+				if err == nil || err == io.EOF {
+					printerId := GetPrinterID(string(line))
+					if printerId != "" {
+						pn := notification.PrinterNotification{printerId, notification.PrinterNewJobs}
+						go func() {
+							fcmNotifications <- pn
```

Comment:
in the xmpp file, they don't wrap the sending of printer notifications in another goroutine. What do you think? Seems like this file should match xmpp (meaning either xmpp is wrong, or this is creating an unnecessary goroutine)

Reply:
goroutines are lite weight, it is better not block fcm listener while publishing to fcmNotifications channel . For lower loads(less notifications per minute) goroutine may not be required but better keep scalability in mind.

Reply:
i think i mentioned this in your initial email why we have a new notification package since this package is essentially only 12 lines...

Reply:
@latin if I understood it right, adlr is asking why we have 

go func() { fcmNotifications <- pn }()          instead of              fcmNotifications <- pn.

And regarding why "notification" package is required :
Notification package holds the common code for both xmpp and fcm notifications.

Reply:
I think it shouldn't be wrapped in another go routine for the following reasons:

- On the receiving side (inside func (pm *PrinterManager) listenNotifications) it already kicks off a new goroutine to do the job
- If we overload printer manager, we can queue up messages in the channel. I don't see why it's preferred to queue up go routes in the scheduler, since even if they are light weight, they can't be that light weight.

Reply:
Fixed.